pyanaconda/modules/subscription/subscription.py

```diff
@@ -302,68 +296,42 @@ def set_system_purpose_data(self, system_purpose_data):
         :param system_purpose_data: system purpose data structure to be set
         :type system_purpose_data: DBus structure
         """
+        # record if the new data is different from current data
+        data_different = self._system_purpose_data != system_purpose_data
+        # set the data
         self._system_purpose_data = system_purpose_data
+        # emit the signal & show log message
         self.system_purpose_data_changed.emit()
         log.debug("System purpose data set to %s.", system_purpose_data)
-
-    @property
-    def is_system_purpose_applied(self):
-        """Report if system purpose has been applied to the system.
-
-        Note that we don't differentiate between the installation environment
-        and the target system, as the token transfer installation task will
-        make sure any system purpose configuration file created in the installation
-        environment will be transferred to the target system.
-
-        We also need to avoid running system purpose configuration again after
-        a successful subscription attempt, as subscription can actually change
-        the system purpose attached to the system via system purpose values
-        attached to an activation key. If we re-run the system purpose task
-        on the installed system, we would basically overwrite these changes.
-        """
-        return self._is_system_purpose_applied
-
-    def set_is_system_purpose_applied(self, system_purpose_applied):
-        """Set if system purpose is applied.
-
-        :param bool system_purpose_applied: True if applied, False otherwise
-
-        NOTE: We keep this as a private method, called by the completed signal of the
-              task that applies system purpose information on the system.
-        """
-        self._is_system_purpose_applied = system_purpose_applied
-        self.is_system_purpose_applied_changed.emit()
-        # as there is no public setter in the DBus API, we need to emit
-        # the properties changed signal here manually
-        self.module_properties_changed.emit()
-        log.debug("System purpose is applied set to: %s", system_purpose_applied)
-
-    def _apply_syspurpose(self):
+        # If new structure contains some data and the data is different than
+        # what was in the old structure, set system purpose data to the system.
+        #
+        # NOTE: We do this only after setting the value and logging the "set"
+        #       as having a set -> apply sequence in the logs makes more sense
+        #       than the other way around.
+        data_available = system_purpose_data.check_data_available()
```

Comment:
I don't get it. If the client sets the system purpose data with no data available, we will set the property, but not the system. Therefore, the system will use the previous configuration, that is now different from our property.

Reply:
This is due to the fact that you need at least one of the values to call the `syspurpose` utility - with no data, you can't really call it. But indeed, like this it is not possibly to clear previously set data, which I think should be possible - eq. the user changes their mind, you want to clear something set by kickstart.

So I'll remove the data available check here & in the task & have the task remove existing syspurpose.json if it gets empty `SystemPurposeData`. Does that sound right ? :)

Reply:
I guess we should always remove the previous configuration. Based on the code, if I set `role` and nothing else, then the system will use the previous configuration for `sla`, `usage` and `addons`.

Reply:
That's a very good point - seems to illustrate a bit that the syspurpose tool was made for simpler usecases that what we have in Anaconda. :)

Reply:
Changed `give_system_purpose()` function to remove existing `syspurpose.json` before running the `syspurpose` tool.